### DISCUSSION TITLE
Question about second parameter of a rule

### DISCUSSION THREAD

**Discussion by forbisc** [tree-sitter/tree-sitter#1917] [Q&A] [answered]
I see rules like this in the tree-sitter-typescript repository and I keep seeing this pattern but I'm not sure what it means.  What does the second parameter `previous` mean and what does placing it as an element of the `choice` do?

```
      declaration: ($, previous) => choice(
        previous,
        $.function_signature,
        $.abstract_class_declaration,
        $.module,
        prec('declaration', $.internal_module),
        $.type_alias_declaration,
        $.enum_declaration,
        $.interface_declaration,
        $.import_alias,
        $.ambient_declaration
      ),
```

EDIT: I guess this is maybe specific to tree-sitter-typescript and tree-sitter-javascript .  Best guess is that previous is referencing existing javascript rules which are extended in the typescript grammar.  Still not totally sure how that works but I think I understand the gist of it.

**Comment by ObserverOfTime** [comment, marked as answer]
previous is, in fact, referencing existing javascript rules which are extended in the typescript grammar.
